### PR TITLE
T11431 bisection docker

### PR DESCRIPTION
**Pull request by gctucker**
Update `bisect.jpl` to run in a Docker image, and update Dodkerfiles to add the required Python packages to run a bisection.

I've built and pushed the resulting images here:
https://hub.docker.com/u/gtucker/

They are a bit larger, for example the `gcc-7_arm64` one is now 230 MB whereas it was 174 MB before.  As these images should not be updated too often and kept in the local cache of the Jenkins builders, this is probably not a big issue.  The advantage of this approach is that the Dockerfiles are easy to manage, with a simple inheritance scheme and no need for docker-compose or anything special.

### CODE REVIEW

**Comment by mattface** [comment]
This does present an interesting new problem, as the bisect builds should really use the docker container that the original builds did, rather than trying to install all the compilers into this 1 image.
This PR will work for the current gcc-7 only setup, but I'm hoping to turn on multiple compilers per build soon and then it gets messy.

We could add the python libs you need to the builder images? Or make new bisection images which extend the builder images plus the python parts?

**Comment by gctucker** [comment]
What I'm planning to do next is to use the arch + compiler type + compiler version to call update-alternatives and pick the matching compiler toolchain in the bisection job.  If it's not available in the image, then it won't run, but that's fine.

I initially thought the same, we should try to build kernels for the bisections with the same image as for the initial builds.  However this is not straightforward for the following reasons:
* the docker image name is not stored in the build meta-data
* we could add it, but it seems fragile to rely on a docker image name when the actual meta-data is the compiler toolchain version
* the docker images to build kernels don't have the python modules required to run LAVA jobs or potentially other things needed for bisection
* running the whole bisection job in a single docker image is a lot cleaner, both from a job coding pov and from a resource usage as it just runs everything on one node

So I think this PR is a first step towards running bisections in Docker containers, but something will need to be done to ensure bisections are run using a toolchain compatible with the one used for the builds.

Thought: It would seem like having a Jenkins library method with a reliable way of mapping "arch + compiler type + compiler version" to a toolchain should solve the problem.  For regular builds, it would be used to work out the name of the docker image to pull, and for bisections the Debian alternative to use.

**Comment by gctucker** [comment]
If we can make a collection of bisection docker images which extend the ones to build kernels in a way that's maintainable and scales, then that's another option.  The same principle of having a way to determine the image to pull still applies, with this time a slightly different image name for bisection jobs.

**Comment by gctucker** [comment]
We could also use "docker compose" to pull an image with the toolchain, one which can run LAVA jobs, and potentially another one with extra packages needed for bisection (in practice I think this might all be covered by the LAVA one).  There's support for that in Jenkins pipeline apparently, I'll investigate.

**Comment by gctucker** [comment]
In the last version, I'm reshaping the dockerfiles slightly to consolidate them and also include the Python packages needed to run bisections and builds with the same images.  So it's easy to manage, but also means slightly larger images, which I don't believe to be an issue.

File: jenkins/bisect.jpl [hunk 1]
```diff
@@ -424,13 +424,110 @@ def checkAbort(passed, message) {
...
+    }
+    if (!check)
+        echo "Warning: revert check failed"
+}
+
+node("docker && bisection") {
     /* Global pipeline constants */
     env._BUILD_JSON = "build-data.json"
 
+    def docker_image = "${params.DOCKER_BASE}${params.COMPILER}_${params.ARCH}"
```
Comment (khilman) [inline jenkins/bisect.jpl]: I'm not sure I fully understand how the docker images are found/created, but this doesn't look quite right with the new hierarchy of Dockerfiles.

**Comment by khilman** [comment]
`$ docker pull kernelci/build-gcc_7_arm
Using default tag: latest
Error response from daemon: pull access denied for kernelci/build-gcc_7_arm, repository does not exist or may require 'docker login'
`

**Comment by gctucker** [comment]
I don't have push access to the kernelci org on Docker hub, so I've tested this with a local build-base image and tags starting with `gtucker/kernelci-build-` as this part of the Docker images can be configured in the Jenkins jobs (DOCKER_BASE parameter):

```
docker build build-base -t kernelci/build-base
for c in gcc-*
  do echo $c
  tag=gtucker/kernelci-build-$c
  echo [$tag]
  docker build $c -t $tag && docker push $tag
done
```

**Pull request merged**